java7/lib/src/main/java/com/truelayer/signing/Verifier.java

```diff
@@ -0,0 +1,263 @@
+     * @return the Verifier instance
+     * @throws KeyException             it the provided key is invalid
+     * @throws IllegalArgumentException if the provided param is null
+     */
+    public static Verifier from(String publicKeyPem) throws JOSEException {
+        if (publicKeyPem == null)
+            throw new IllegalArgumentException("the publicKey must not be null");
+
+        ECPublicKey publicKey = ECKey.parseFromPEMEncodedObjects(publicKeyPem).toECKey().toECPublicKey();
+        return new VerifierFromPublicKey(publicKey);
+    }
+
+    /**
+     * Start building a `Tl-Signature` header verifier using public key RFC 7468 PEM-encoded data.
+     *
+     * @param publicKeyPem the public key 7468 PEM-encoded data - must not be null
+     * @return the Verifier instance
+     * @throws KeyException             it the provided key is invalid
+     * @throws IllegalArgumentException if the provided param is null
+     */
+    public static Verifier from(ECPublicKey publicKeyPem) {
+        if (publicKeyPem == null)
+            throw new IllegalArgumentException("the publicKey must not be null");
+
+        return new VerifierFromPublicKey(publicKeyPem);
+    }
+
+    /**
+     * Start building a `Tl-Signature` header verifier using public key JWKs JSON response data.
+     *
+     * @param jwks public key JWKs JSON response data
+     * @return the Verifier instance
+     * @throws SignatureException if the provided jwks is invalid
+     */
+    public static Verifier verifyWithJwks(String jwks) throws ParseException {
+        JWKSet jwkSet = JWKSet.parse(jwks);
+        return new VerifierFromJwks(jwkSet);
+    }
+
+    /**
+     * Verify the given `Tl-Signature` header value.
+     *
+     * @param signature the given `TL-signature`
+     * @throws SignatureException if Signature is invalid
+     */
+    public abstract void verify(String signature);
+
+    protected Map<HeaderName, String> validateSignatureHeader(JWSHeader jwsHeader) {
+
+        List<String> validSignatureHeaders = validateSignatureHeaders(jwsHeader, requiredHeaders);
+
+        LinkedHashMap<HeaderName, String> orderedHeaders = new LinkedHashMap<>();
+
+        for (String header: validSignatureHeaders) {
+            String val = headers.get(new HeaderName(header));
+            if (val != null)
+                orderedHeaders.put(new HeaderName(header), val);
+        }
+
+        return orderedHeaders;
+    }
+
+    private List<String> validateSignatureHeaders(JWSHeader jwsHeader, HashSet<String> requiredHeaders) throws SignatureException {
+        SignatureException.ensure(jwsHeader.getAlgorithm().equals(JWSAlgorithm.ES512), "unsupported jws alg");
+        SignatureException.ensure(jwsHeader.getCustomParam("tl_version").toString().equals("2"), "unsupported jws tl_version");
+
+        List<String> signedHeaders = Arrays.asList(jwsHeader.getCustomParam("tl_headers").toString().split(","));
+
+        List<String> missingHeaders = new ArrayList<>();
```

Comment:
as an alternative. We could check third party libs like Guava or Apache Common collection like mentioned on this article. As long as unit tests are green I would not bother though 👍